src/test/groovy/com/wooga/gradle/extensions/ProviderExtensionsSpec.groovy

```diff
@@ -0,0 +1,201 @@
+                { p -> p.map { "otherValue" }.get() },
+        ]
+        baseException = new Exception("base exception")
+        message = "my error"
+    }
+
+    def "runs mapOnce function only once on multiple provider executions"() {
+        given:
+        def execCounter = 0
+        def provider = project.provider { "a" }.mapOnce { execCounter++ }
+        when:
+        provider.get()
+        provider.get()
+        then:
+        execCounter == 1
+    }
+
+    def "mapOnce function always returns same value as the first run"() {
+        given:
+        def execCounter = 0
+        def provider = project.provider { "a" }.mapOnce { ++execCounter }
+        def otherProvider = project.provider { "a" }.map { ++execCounter }
+
+        when:
+        def result = provider.get()
+        def secondResult = provider.get()
+        def otherResult = otherProvider.get()
+        def otherSecondResult = otherProvider.get()
+
+        then:
+        result == secondResult
+        otherResult != otherSecondResult
+    }
+
+    @Unroll("logs #level-level #situation message once on provider resolution")
+    def "logs message once on provider resolution"() {
+        given:
+        def logger = Mock(Logger)
+        def provider = project.provider { value }
+        switch (level) {
+            case "warn": provider = provider.logWarn(logger, message); break
+            case "info": provider = provider.logInfo(logger, message); break
+            case "debug": provider = provider.logDebug(logger, message); break
+            case "error": provider = provider.logError(logger, message); break
+        }
+
+        when:
+        provider.get()
+        provider.get()
+
+        then:
+        switch (level) {
+            case "warn": 1 * logger.warn(expected); break
+            case "info": 1 * logger.info(expected); break
+            case "debug": 1 * logger.debug(expected); break
+            case "error": 1 * logger.error(expected); break
+        }
+
+        where:
+        level   | message             | value   | expected
+        "warn"  | "msg"               | "value" | message
+        "warn"  | { it -> "$it msg" } | "value" | "$value msg"
+        "info"  | "msg"               | "value" | message
+        "info"  | { it -> "$it msg" } | "value" | "$value msg"
+        "debug" | "msg"               | "value" | message
+        "debug" | { it -> "$it msg" } | "value" | "$value msg"
+        "error" | "msg"               | "value" | message
+        "error" | { it -> "$it msg" } | "value" | "$value msg"
+        situation = message instanceof String ? "static" : "enclosed"
+    }
+
+    def "error level logger can log exception once"() {
+        given:
+        def logger = Mock(Logger)
+        def provider = project.provider {""}.logError(logger, message, exception)
+
+        when:
+        provider.get()
+        provider.get()
+
+        then:
+        1 * logger.error(message, exception)
```

Comment:
Can you elaborate on this?

Reply:
Logger is a Mock, so this mean that logger.error with those parameters should be executed exactly once.

Reply:
https://spockframework.org/spock/docs/1.0/interaction_based_testing.html

Reply:
Oh I understand that. I was thinking of writing a small comment on the invocation since it's unconventional but I guess it's not needed.